library/pod-security-policy/users/template.yaml

```diff
@@ -132,158 +132,640 @@ spec:
+          - name: podRunAsGroup
+            expression: |
+              variables.anyObject.kind == "Pod" && has(variables.anyObject.spec.securityContext) && has(variables.anyObject.spec.securityContext.runAsGroup) ? variables.anyObject.spec.securityContext.runAsGroup : null
+          - name: podRunAsFsGroup
+            expression: |
+              variables.anyObject.kind == "Pod" && has(variables.anyObject.spec.securityContext) && has(variables.anyObject.spec.securityContext.fsGroup) ? variables.anyObject.spec.securityContext.fsGroup : null
+          - name: missingRunAsNonRootGlobal
+            expression: |
+              !has(variables.anyObject.securityContext) || ((!has(variables.anyObject.securityContext.runAsNonRoot) || !variables.anyObject.securityContext.runAsNonRoot) && (!has(variables.anyObject.securityContext.runAsUser) || variables.anyObject.securityContext.runAsUser == 0))
+          - name: missingRequiredRunAsUserContainers
+            expression: |
+              variables.containers.filter(container, 
+                !(container.image in variables.exemptImages) &&
+                has(variables.params.runAsUser) && has(variables.params.runAsUser.rule) && (variables.params.runAsUser.rule == "MustRunAs") &&
+                !((has(container.securityContext) && has(container.securityContext.runAsUser)) || 
+                  (variables.podRunAsUser != null)))
+          - name: missingRequiredRunAsNonRootContainers
+            expression: |
+              variables.containers.filter(container, 
+                !(container.image in variables.exemptImages) &&
+                has(variables.params.runAsUser) && has(variables.params.runAsUser.rule) && (variables.params.runAsUser.rule == "MustRunAsNonRoot") ?
+                (
+                  (!has(container.securityContext) || (
+                    (!has(container.securityContext.runAsNonRoot) || !container.securityContext.runAsNonRoot) && (!has(container.securityContext.runAsUser) || container.securityContext.runAsUser == 0) 
+                  )) || variables.missingRunAsNonRootGlobal
+                ) : false
+              )
+          - name: processedRunAsUserContainers
+            expression: (variables.missingRequiredRunAsNonRootContainers + variables.missingRequiredRunAsUserContainers).map(container, container.name)
+          - name: invalidRunAsUserContainers
+            expression: |
+              variables.containers.filter(container, 
+                !(container.image in variables.exemptImages) &&
+                !(container.name in variables.processedRunAsUserContainers) &&
+                has(variables.params.runAsUser) && has(variables.params.runAsUser.rule) ? 
+                (
+                  variables.params.runAsUser.rule == "RunAsAny" ? false :
+                  (
+                    variables.params.runAsUser.rule == "MustRunAsNonRoot" ?
```

Comment:
I think the code for this will be much simpler if we:

* have a variable for each rule type (`MustRunAs`, `MayRunAs`, etc.) that returns the containers violating the rule.
* if the rule type is not used, this variable will return an empty list.

This will allow you to:

* remove all code duplication (can go back to concatenating all containers again, since you are no longer required to filter based off container name)
* Reduce the nesting/branching for each variable.

Because this will significantly change/shorten the code, I'll stop reviewing here.

Reply:
Thanks for the suggestion. Made an update to CEL code, it should now be much more compact. PTAL.